src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -4041,10 +4041,15 @@ bool PhaseIdealLoop::intrinsify_fill(IdealLoopTree* lpt) {
     index = new LShiftXNode(index, shift->in(2));
     _igvn.register_new_node_with_optimizer(index);
   }
-  index = new AddPNode(base, base, index);
-  _igvn.register_new_node_with_optimizer(index);
-  Node* from = new AddPNode(base, index, offset);
+  Node* from = new AddPNode(base, base, index);
   _igvn.register_new_node_with_optimizer(from);
+  // For normal array fills, C2 uses two AddP nodes for array element
+  // addressing. But for array fills with Unsafe call, there's only one
+  // AddP node adding an absolute offset, so we do a NULL check here.
+  if (offset != NULL) {
+    from = new AddPNode(base, from, offset);
```

Comment:
Is it worth adding an `C->has_unsafe_access()` assert here?

Reply:
An assertion is added in new commit. Thanks for your review.